### PR TITLE
feat: add Brave search tools with example

### PR DESCRIPTION
**Pull request by im-vedant**
This PR adds Brave search tools with an example implementation.

## Changes

- Add Brave search integration with new tools
- Include example usage and implementation
- **Breaking Change**: Web search tool has been migrated from `web_search` to `duckduckgosearch` tool
- Parameter naming change: `query` parameter has been updated to `webAction`

## Documentation & Migration Notes

⚠️ **Action Required**: Existing prompts referencing `web_search` or the `query` parameter will need to be updated:
- Update tool name from `web_search` to `duckduckgosearch`
- Update parameter from `query` to `webAction`

This change should be documented in release notes to ensure users can migrate their integrations accordingly.

### CODE REVIEW

File: modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/BraveSearchTool.scala [hunk 1]
```diff
@@ -0,0 +1,319 @@
+package org.llm4s.toolapi.builtin.search
+
+import org.llm4s.toolapi._
+import upickle.default._
+import org.llm4s.config.Llm4sConfig
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/BraveSearchTool.scala]: It imports org.llm4s.config.Llm4sConfig and calls Llm4sConfig.loadBraveSearchTool() inside the tool handler and inside withApiKey.
But standardization rule is, the core code should not read config or depend on Llm4sConfig. Config loading must stay in org.llm4s.config, and the app layer (samples/CLI/tests) should load it and pass it in to the core code in dependency injection style.

File: modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/BraveSearchTool.scala [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+  private def search[R](
+    query: String,
+    config: BraveSearchConfig,
+    braveTool: BraveTool,
+    category: BraveSearchCategory[R]
+  ): Either[String, R] = {
+    import sttp.client4._
+
+    Try {
+      val backend = DefaultSyncBackend()
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/BraveSearchTool.scala]: create DefaultSyncBackend() per request and never close it. Wrap in
Using.resource(DefaultSyncBackend()) (or inject/reuse a backend) to avoid leaking sockets/threads

File: modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/DuckDuckGoSearchTool.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  val SAFE_SEARCH                                                  = "1"
+  val UNSAFE_SEARCH                                                = "-1"
+  private def search(
+    query: String,
+    config: DuckDuckGoSearchConfig
+  ): Either[String, DuckDuckGoSearchResult] = {
+    import sttp.client4._
+
+    Try {
+      val backend = DefaultSyncBackend()
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/DuckDuckGoSearchTool.scala]: create DefaultSyncBackend() per request and never close it. Wrap in
Using.resource(DefaultSyncBackend()) (or inject/reuse a backend) to avoid leaking sockets/threads

File: modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/DuckDuckGoSearchTool.scala [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            abstractSource = json.obj.get("AbstractSource").map(_.str).getOrElse(""),
+            abstractUrl = json.obj.get("AbstractURL").map(_.str).getOrElse(""),
+            answer = json.obj.get("Answer").map(_.str).getOrElse(""),
+            answerType = json.obj.get("AnswerType").map(_.str).getOrElse(""),
+            relatedTopics = relatedTopics,
+            infoboxContent = infobox
+          )
+        case Left(error) =>
+          throw new Exception(s"HTTP request failed: $error")
+      }
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/DuckDuckGoSearchTool.scala]: inside Try. Prefer returning Left(errorMsg) directly (no throw) so failure paths stay explicit and total.

File: modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/BraveSearchTool.scala [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+        .send(backend)
+
+      response.body match {
+        case Right(body) =>
+          // Parse the JSON response and extract the first result
+          val json = ujson.read(body)
+          category.parseResults(json, query)
+        case Left(error) =>
+          throw new Exception(s"HTTP request failed: $error")
+      }
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/BraveSearchTool.scala]: inside Try. Prefer returning Left(errorMsg) directly (no throw) so failure paths stay explicit and total.

File: modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/package.scala [hunk 6]
```diff
@@ -3,31 +3,32 @@ package org.llm4s.toolapi.builtin
...
- * val searchTool = WebSearchTool.create()
+ * val searchTool = DuckDuckGoSearchTool.tool
  *
- * // Custom configuration
- * val customSearch = WebSearchTool.create(WebSearchConfig(
- *   timeoutMs = 15000,
- *   maxResults = 5
- * ))
+ * // Brave search tool
+ * val braveSearch = BraveSearchTool.tool
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/package.scala]: refers to BraveSearchTool.tool, but BraveSearchTool do not have a tool member.

File: modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/BraveSearchTool.scala [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    case "strict"   => Strict
+    case _          => Moderate // Safe default
+  }
+}
+
+case class BraveSearchConfig(
+  count: Int = 5,
+  safeSearch: SafeSearch = SafeSearch.Strict,
+  extraParams: Map[String, Any] = Map.empty
+)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/toolapi/builtin/search/BraveSearchTool.scala:142]: Using Any here makes the code more loose. It hides mistakes until runtime - violates tyep safety.

Prefer something typed like Map[String, String] - or, if you truly need mixed values, use ujson.Obj so the values still have real JSON types instead of Any

File: modules/core/src/main/scala/org/llm4s/config/ToolsConfigLoader.scala
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/config/ToolsConfigLoader.scala:1]: This class / file is in the correct package to use PureConfig, but tools should accept BraveTool/config explicitly; samples can call Llm4sConfig.loadBraveSearchTool() and pass the result into tool construction.

**Comment by rorygraves** [comment]
## Code Review: PR #391 - Brave Search Tools

Thanks for the updates @im-vedant! I've reviewed the latest code against our [review guidelines](https://github.com/llm4s/llm4s/blob/main/docs/reference/review-guidelines.md).

---

### ✅ What's Good

| Guideline | Status | Notes |
|-----------|--------|-------|
| **#2 Configuration - DI** | ✅ Fixed | `BraveSearchTool.create(toolConfig, ...)` accepts config as parameter |
| **#7 JSON Handling** | ✅ Good | Uses ujson with safe `.strOpt`, `.objOpt`, `.arrOpt` patterns |
| **#1 Error Handling** | ✅ Good | Uses `Result[A]` for config loading, `Either` for tool operations |
| **#4 HTTP Client** | ✅ Good | Uses `requests` library (Java HttpClient) per project standards |
| **Testing** | ✅ Good | Comprehensive unit tests for both tools |
| **Sealed Traits** | ✅ Good | Clean `BraveSearchCategory` and `SafeSearch` ADTs |

---

### 🔴 Issues to Address

#### 1. Debug Statement Left in Code
**Location:** `BraveSearchTool.scala:480`
```scala
print("Some error occurred")  // ❌ Remove or use proper logging
```

---

#### 2. Config Loading in Object Initialization (Guideline #5)
**Location:** `ResearcherAgentExample.scala:1431`
```scala
private val config         = ConfigFactory.load()
private val RESEARCH_TOPIC = config.getString("llm4s.samples.agent.research-topic")
```
This loads config during object initialization (side effect). Move to `main()`:
```scala
def main(args: Array[String]): Unit = {
  val config = ConfigFactory.load()
  val researchTopic = config.getString("llm4s.samples.agent.research-topic")
  // ...
}
```

---

#### 3. Breaking API Changes Need Documentation (Guideline #14)
Two breaking changes not documented in the PR:
- Tool name: `web_search` → `duckduckgo_search`
- Parameter name: `query` → `search_query`

**Action:** Add to PR description/release notes that existing prompts referencing `web_search` or `query` parameter will need updates.

---

#### 4. PR Size (Guideline: Keep PRs Small)
This PR combines multiple concerns:
1. Renaming existing DuckDuckGo tool
2. Adding new Brave Search tools
3. Adding config infrastructure (`ToolsConfigLoader`)
4. Adding researcher agent sample

Consider splitting into smaller focused PRs, or at minimum document the scope clearly in the PR description.

---

#### 5. logback.xml May Override User Config
**Location:** `modules/samples/src/main/resources/logback.xml`

Adding this to samples could unexpectedly override logging configuration in applications that include samples as a dependency. Consider moving to `src/main/resources-filtered/` or documenting this behavior.

---

### 🟡 Minor Suggestions

#### 6. Additional Test Cases
Consider adding failure case tests:
- Network timeout handling
- Invalid API responses
- Empty search results edge cases

---

### Summary

| Priority | Issue | Action |
|----------|-------|--------|
| 🔴 Critical | Debug `print` statement | Remove line 480 |
| 🔴 Important | Config in object init | Move to `main()` |
| 🟠 Important | Breaking changes undocumented | Add to PR description |
| 🟠 Suggestion | Large PR size | Consider splitting or documenting scope |
| 🟡 Minor | logback.xml override risk | Document or relocate |

The PR addresses @vim89's DI feedback well. Main remaining issues are the debug statement and config loading pattern. Once those are fixed, this should be ready to merge!

**Comment by rorygraves** [comment]
@im-vedant Great work, posted an updated review summary above.

**Comment by rorygraves** [comment]
Thanks @im-vedant!
